protostar-rust/src/lib.rs

```diff
@@ -1,36 +1,71 @@
-use pyo3::exceptions::RuntimeError;
-use pyo3::prelude::*;
-use pyo3::wrap_pyfunction;
-
-use anyhow::Context;
-use std::collections::HashMap;
-
-use cairo_lang_protostar::test_collector::collect_tests;
+use anyhow::{Context, Result};
+use cairo_lang_protostar::test_collector::{collect_tests, LinkedLibrary};
 use cairo_lang_runner::{RunResultValue, SierraCasmRunner};
+use serde::Deserialize;
+use std::path::PathBuf;
+use walkdir::WalkDir;
 
 fn run_result_value_to_string(run_result: RunResultValue) -> String {
-    match run_result {
-        RunResultValue::Success(data) => {
-            return format!("Success {:?}", data);
+    return match run_result {
+        RunResultValue::Success(data) => format!("Success {:?}", data),
+        RunResultValue::Panic(data) => format!("Panic {:?}", data),
+    };
+}
+
+fn collect_tests_in_directory(input_path: &PathBuf) -> Vec<PathBuf> {
+    let mut test_directories: Vec<PathBuf> = vec![];
+
+    for entry in WalkDir::new(input_path) {
+        if entry.is_err() {
+            continue;
         }
-        RunResultValue::Panic(data) => {
-            return format!("Panic {:?}", data);
+
+        let entry = entry.unwrap();
+        let path = entry.path();
+
+        if path.is_file() && path.extension().map_or(false, |ex| ex == "cairo") {
+            test_directories.push(path.to_path_buf());
         }
     }
+
+    test_directories
 }
 
-fn internal_run_tests(input_path: &str) -> anyhow::Result<()> {
-    let (sierra_program, test_configs) = collect_tests(&input_path.to_owned(), None, None, None)?;
+pub fn run_tests(
+    input_path: PathBuf,
+    linked_libraries: Option<Vec<LinkedLibrary>>,
+    config: &ProtostarTestConfig,
+) -> Result<()> {
+    let test_directories = collect_tests_in_directory(&input_path);
+    for test in test_directories {
+        run_tests_in_file(test, linked_libraries.clone())?;
+    }
+    Ok(())
+}
+
+fn run_tests_in_file(
+    input_path: PathBuf,
+    linked_libraries: Option<Vec<LinkedLibrary>>,
+) -> Result<()> {
+    let builtins = vec!["GasBuiltin", "Pedersen", "RangeCheck", "bitwise", "ec_op"];
```

Comment:
wouldn't it be better to define this as static or something? Logically, this will allocate memory multiple times, if we do this as `static` we allocate just one time, right? And it can be evaluated in the compile time.

Reply:
I think compiler APIs expect a Vec here

Reply:
We could define a static array of these but we'd have to construct a vector from these anytime we call `collect_tests` anyway. Don't know if that would be better at all.

cc @mkaput

Reply:
The `vec!` macro basically does what you said @cptartur